fairscale/nn/misc/auto_wrap.py

```diff
@@ -0,0 +1,148 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch.nn as nn
+import contextlib
+from typing import Generator, Union, Tuple, Callable
+
+from fairscale.nn.data_parallel.fully_sharded_data_parallel import FullyShardedDataParallel
+
+
+@contextlib.contextmanager
+def enable_wrap(**kwargs) -> Generator[None, None, None]:
+    """
+    Context manager to wrap modules in FullyShardedDataParallel.
+
+    Useful for when you'd like to apply the same parameters to all child modules
+    that you wrap. A particularly important use case is wrapping large layers so
+    that they get sharded (in-place) during initialization, to avoid running out of
+    system memory. Large layers can indicate that they should be sharded via
+    the ``wrap`` annotation and this context manager can provide the
+    exact configuration for these nested instances.
+
+    Usage::
+
+            with enable_wrap(**params):
+                # Wraps layer in FSDP by default if within context
+                self.l1 = wrap(torch.nn.Linear(5, 5))
+                # Wraps children modules by default based on min_num_params
+                self.l2 = auto_wrap(TransformerBlock(), min_num_params=1e8)
+
+    """
+    with ConfigAutoWrap(**kwargs):
+        yield
+
+
+def wrap(module, cls: Callable = FullyShardedDataParallel, **wrap_overrides):
+    """
+    Annotate that a module should be wrapped.
+    Annotated modules will only be wrapped if inside of an ``enable_wrap``
+    context manager. An important use case is annotating large layers that
+    should be sharded (in-place) during initialization, to avoid running out
+    of system memory.
+
+    Usage::
+
+        with enable_wrap(**params):
+            # Wraps layer in FSDP by default if within context
+            self.l1 = wrap(torch.nn.Linear(5, 5))
+
+    Args:
+        module (nn.Module): module to wrap (if in ``enable_wrap`` context)
+        cls (Callable): Class wrapper to wrap the model with if in context (Default FullyShardedDataParallel)
+        **wrap_overrides: Configuration overrides that will take
+            priority over the values provided by the ``enable_wrap`` context
+    """
+    if ConfigAutoWrap.in_autowrap_context:
+        wrap_overrides = {**ConfigAutoWrap.kwargs, **wrap_overrides}
+        return cls(module, **wrap_overrides)
+    return module
+
+
+def auto_wrap(module, min_num_params: float = 1e8, cls: Callable = FullyShardedDataParallel, **kwargs):
```

Comment:
If this works, we're going to need a more sensible default here, I just chose this randomish